common/lib/xmodule/xmodule/modulestore/split_mongo/split.py

```diff
@@ -2045,6 +2044,157 @@ def copy(self, user_id, source_course, destination_course, subtree_list=None, bl
+
+        destination_course = dest_usage.course_key
+        with self.bulk_operations(destination_course):
+            index_entry = self.get_course_index(destination_course)
+            if index_entry is None:
+                raise ItemNotFoundError(destination_course)
+            dest_structure = self._lookup_course(destination_course).structure
+            old_dest_structure_version = dest_structure['_id']
+            dest_structure = self.version_structure(destination_course, dest_structure, user_id)
+
+            # Set of all descendent block IDs of dest_usage that are to be replaced:
+            block_key = BlockKey(dest_usage.block_type, dest_usage.block_id)
+            orig_descendants = set([block for block in self.descendants(dest_structure['blocks'], block_key, depth=None, descendent_map={})])
+            orig_descendants.remove(block_key)  # The descendants() method used above adds the block itself, which we don't consider a descendant.
+            new_descendants = self._inherit_copy(source_structures, source_keys, dest_structure, block_key, user_id, recurse=copy_children)
+
+            # Update the edit info:
+            dest_info = dest_structure['blocks'][block_key]
+
+            # Update the edit_info:
+            dest_info['edit_info']['previous_version'] = dest_info['edit_info']['update_version']
+            dest_info['edit_info']['update_version'] = old_dest_structure_version
+            dest_info['edit_info']['edited_by'] = user_id
+            dest_info['edit_info']['edited_on'] = datetime.datetime.now(UTC)
+
+            orphans = orig_descendants - new_descendants
+            for orphan in orphans:
+                del dest_structure['blocks'][orphan]
+
+            self.update_structure(destination_course, dest_structure)
+            self._update_head(destination_course, index_entry, destination_course.branch, dest_structure['_id'])
+        # Return usage locators for all the new children:
+        return [destination_course.make_usage_key(*k) for k in dest_structure['blocks'][block_key]['fields']['children']]
+
+    def _inherit_copy(self, source_structures, source_keys, dest_structure, new_parent_block_key, user_id, recurse):
+        """
+        Internal recursive implementation of inherit_copy()
+
+        Returns the new set of BlockKeys that are the new descendants of the block with key 'block_key'
+        """
+        # pylint: disable=no-member
+        # ^-- Until pylint gets namedtuple support, it will give warnings about BlockKey attributes
+        new_blocks = set()
+
+        new_children = list()  # ordered list of the new children of new_parent_block_key
+
+        for usage_key in source_keys:
+            src_course_key = usage_key.course_key.for_version(None)
+            block_key = BlockKey(usage_key.block_type, usage_key.block_id)
+            source_structure = source_structures.get(src_course_key, [])
+            if block_key not in source_structure['blocks']:
```

Comment:
It's improbable, but it's not impossible - because `source_structures` is only generated from the top-level usage keys, but this method recurses to those block's children. If one of those blocks or any of its descendants has an invalid "children" setting that refers to a block missing from that course/library's structure, this could occur. That said, the only way that would happen is if at some point there was an error during some modulestore operation, resulting in the structure having an inconsistent state. (Something that has happened to me while coding changes to split, but shouldn't happen in production.)